### PR TITLE
generate: consider service accounts when generating a CSV

### PR DESCRIPTION
**Pull request by estroz**
**Description of the change:**
*  internal/cmd/operator-sdk/generate: write RBAC objects to stdout or files named with object.Name + GVK, and rename `--update-crds` to `--update-objects`
* internal/generate/{collector/clusterserviceversion}: consider (cluster) role bindings so CSV generator can assign the correct service account names to roles

**Motivation for the change:**  This PR adds handling for extra RBAC objects present in `generate <bundle|packagemanifests>` input. These objects will be written to the resulting bundle. For now, only Roles, RoleBindings, their Cluster equivalents, and ServiceAccounts are written.

This PR also correctly names service account for (cluster) role permissions. These are currently incorrect because the CSV generator is naively using (cluster) role names instead of actual service account names. Previously this was ok because the names match the service account, but this is no longer the case. See #3600.

Old test data has been removed, and a static `basic.operator.yaml` containing the output of `kustomize build config/manifests` added; the static file's contents match a current project manifest build.

**Checklist**

If the pull request includes user-facing changes, extra documentation is required:
- [x] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))
- [x] Add or update relevant sections of the docs website in [`website/content/en/docs`](https://github.com/operator-framework/operator-sdk/tree/master/website/content/en/docs)

### CODE REVIEW

File: internal/generate/clusterserviceversion/clusterserviceversion_updaters.go [hunk 1]
```diff
@@ -80,36 +81,110 @@ func getCSVInstallStrategy(csv *operatorsv1alpha1.ClusterServiceVersion) operato
...
+			if _, ok := roleNames[role.GetName()]; ok {
+				p.Rules = append(p.Rules, role.Rules...)
+			} else {
+				unmappedRoles = append(unmappedRoles, role)
+			}
+		}
+		perms = append(perms, p)
 	}
 	strategy.Permissions = perms
+	// Add all unmapped roles to a permission without a service account.
```
Comment (estroz) [inline internal/generate/clusterserviceversion/clusterserviceversion_updaters.go:116]: This isn't the right thing to do (deployment will fail), but should the generator assume in this case that one of the service accounts in a binding can be used here? Is this always the correct thing to do? The default SDK Go project doesn't scaffold a cluster role binding for the [`metrics-reader`](https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v3/config/rbac/auth_proxy_client_clusterrole.yaml) cluster role.

/cc @joelanford @hasbro17

File: internal/generate/clusterserviceversion/clusterserviceversion_updaters.go [hunk 2]
```diff
@@ -80,36 +81,110 @@ func getCSVInstallStrategy(csv *operatorsv1alpha1.ClusterServiceVersion) operato
...
-	strategy.ClusterPermissions = perms
+	return saNames
 }
 
 // applyDeployments updates strategy's deployments with the Deployments
 // in the collector.
 func applyDeployments(c *collector.Manifests, strategy *operatorsv1alpha1.StrategyDetailsDeployment) {
 	depSpecs := []operatorsv1alpha1.StrategyDeploymentSpec{}
 	for _, dep := range c.Deployments {
+		// TODO(estroz): ensure each deployment's pod template spec contains a service account. There may be more
```
Comment (estroz) [inline internal/generate/clusterserviceversion/clusterserviceversion_updaters.go:185]: The CSV will install successfully without one (tested using an example operator) so this isn't necessary to deal with right away.

File: internal/generate/clusterserviceversion/clusterserviceversion_updaters.go (hunk 1 above)
Comment (joelanford) [reply internal/generate/clusterserviceversion/clusterserviceversion_updaters.go:116]: I haven't grokked this PR yet and may not understand the context for your question. So take the following with a grain of salt if you've already thought about this.

The operator doesn't need `metrics-reader`. There are some extra roles that can be bound post-deploy by an admin to enable client interaction with the operator, like this `metrics-reader` one and also the reader and editor roles for each of the CRDs.

I think for any roles that don't have corresponding bindings, maybe we add them to the manifests directory separately, outside the CSV?

File: internal/generate/clusterserviceversion/clusterserviceversion_updaters.go (hunk 2 above)
Comment (joelanford) [reply internal/generate/clusterserviceversion/clusterserviceversion_updaters.go:185]: I think we just use what's in the deployment. If there isn't one listed, that implicitly means `default`.

File: internal/generate/clusterserviceversion/clusterserviceversion_updaters.go (hunk 1 above)
Comment (estroz) [reply internal/generate/clusterserviceversion/clusterserviceversion_updaters.go:116]: >we add them to the manifests directory separately, outside the CSV

I'm in favor of doing that.

Comment (joelanford) [reply internal/generate/clusterserviceversion/clusterserviceversion_updaters.go:116]: In fact, based on my other comment, I'm kind of thinking that the CSV should only have RBAC rules for the service account used by the deployment.

And all other serviceaccounts, roles, rolebindings, clusterroles, and clusterrolebindings that we slurp in from STDIN are just added to the manifests folder separately.

Comment (estroz) [reply internal/generate/clusterserviceversion/clusterserviceversion_updaters.go]: We should probably warn the user if no service account is specified in a deployment's pod template spec then.

Comment (estroz) [reply internal/generate/clusterserviceversion/clusterserviceversion_updaters.go]: I think for now mapping (cluster) roles to service accounts using role bindings is ok. When we can patch a deployment to have a pod service account by default, we can change the internal implementation. Thoughts @joelanford?

Comment (joelanford) [reply internal/generate/clusterserviceversion/clusterserviceversion_updaters.go]: I think its fairly common to not specify `serviceAccountName` in pod specs. And I think that's totally okay. 

If we see that `spec.serviceAccountName == ""`, that's equivalent to `spec.serviceAccountName == "default"` so when we're looking at role bindings, we look for service account "default" when the spec service account is empty.

**Comment by estroz** [comment]
The algorithm for writing permissions, as a result of discussion in https://github.com/operator-framework/operator-sdk/pull/3610#discussion_r462690659 and with @joelanford offline:
1. for each `Deployment` in input:
	1. set `operatorSA` to `.spec....serviceAccountName`, or to "default" if empty
	1. for each `(Cluster)Role` in the input:
		1. find a referent `(Cluster)RoleBinding`
		1. if a referent does not exist, write the `(Cluster)Role` to the manifests directory
		1. if a referent does exist:
			1. if it binds to `operatorSA`, add the `(Cluster)Role` rules to the CSV (cluster)permissions for `operatorSA`
			1. if it does not bind to `operatorSA`, write the referent `(Cluster)RoleBinding` and the referenced `(Cluster)Role` to the manifests directory

This algorithm is split into two parts:
1. `collector.Manifests.GetNonCSVObjects()` will find all `(Cluster)Role`s unbound and bound but not to a CSV-relevant service account and write them to the manifests directory.
1. The CSV generator will add all rules of `(Cluster)Role` bound to a deployment's service account to the correct set of permissions.

File: internal/generate/clusterserviceversion/clusterserviceversion_test.go [hunk 3]
```diff
@@ -45,16 +43,12 @@ var (
...
-	goTestDataDir = filepath.Join(testDataDir, "non-standard-layout")
-	goAPIsDir     = filepath.Join(goTestDataDir, "api")
-	goConfigDir   = filepath.Join(goTestDataDir, "config")
-	goCRDsDir     = filepath.Join(goConfigDir, "crds")
+	csvNewLayoutBundleDir = filepath.Join(csvDir, "output")
+
+	goTestDataDir       = filepath.Join(testDataDir, "go")
+	goAPIsDir           = filepath.Join(goTestDataDir, "api")
+	goStaticDir         = filepath.Join(goTestDataDir, "static")
+	goBasicOperatorPath = filepath.Join(goStaticDir, "basic.operatory.yaml")
```
Comment (joelanford) [inline internal/generate/clusterserviceversion/clusterserviceversion_test.go:51]: Nit, typo:
```suggestion
	goBasicOperatorPath = filepath.Join(goStaticDir, "basic.operator.yaml")
```

File: internal/generate/testdata/clusterserviceversions/output/memcached-operator.clusterserviceversion.yaml [hunk 4]
```diff
@@ -8,10 +8,10 @@ metadata:
           "apiVersion": "cache.example.com/v1alpha1",
```
Comment (estroz) [inline internal/generate/testdata/clusterserviceversions/output/memcached-operator.clusterserviceversion.yaml:8]: These testdata changes can be refactored into another PR, they're not directly related to the bugfix.
